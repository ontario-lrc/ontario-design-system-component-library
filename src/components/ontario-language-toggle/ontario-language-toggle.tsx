import { Component, Prop, State, Event, EventEmitter, Watch, h, Fragment } from '@stencil/core';

import { Language } from '../../utils/common/language-types';
import { validateLanguage } from '../../utils/validation/validation-functions';

import { default as translations } from '../../translations/global.i18n.json';
import { HeaderLanguageToggleEventDetails } from '../../utils/events/common-events.interface';

@Component({
	tag: 'ontario-language-toggle',
	styleUrl: 'ontario-language-toggle.scss',
	shadow: true,
})
export class OntarioLanguageToggle {
	@Prop({ mutable: true }) language: Language | string;

	/**
	 * The size of the language toggle button.
	 *
	 * If no prop is passed, it will be set to the `default` size.
	 */
	@Prop() size?: 'default' | 'small' = 'default';

	/**
	 * The URL to change to when the language toggle button is clicked.
	 *
	 * This is optional.
	 */
	@Prop({ mutable: true }) url?: string;

	/**
	 * A custom function to pass to the language toggle button.
	 *
	 * This is optional.
	 */
	@Prop() customLanguageToggle?: (event: globalThis.Event) => void;

	@State() translations: any = translations;

	/**
	 * An event to set the Document's HTML lang property, and emit the toggled language to other components.
	 */
	@Event() setAppLanguage: EventEmitter<string>;
	setAppLanguageHandler() {
		let lang: string | Language;
		if (this.language) {
			lang = this.language;
		} else if (document.documentElement.lang) {
			lang = document.documentElement.lang;
		} else {
			lang = 'en';
		}

		this.language = lang;
		this.setAppLanguage.emit(lang);

		this.updateHTMLLang(lang);
	}

	/**
	 * An event that emits to other components that the language toggle button has been toggled.
	 */
	@Event() headerLanguageToggled: EventEmitter<HeaderLanguageToggleEventDetails>;
	handleHeaderLanguageToggled(language: string, event?: globalThis.Event) {
		const toggledLanguage = language === 'en' ? 'fr' : 'en';
		this.language = toggledLanguage;

		this.headerLanguageToggled.emit({ oldLanguage: language, newLanguage: toggledLanguage });

		this.updateHTMLLang(toggledLanguage);

		if (this.customLanguageToggle && event) {
			this.customLanguageToggle(event);
		}
	}

	@Watch('language')
	updateLanguage() {
		this.language = validateLanguage(this.language);
		this.setAppLanguageHandler();
	}

	updateHTMLLang = (lang: string) => {
		const htmlElement = document.firstElementChild;

		if (htmlElement?.tagName.toLowerCase() === 'html') {
			if (lang) {
				htmlElement.setAttribute('lang', lang);
			} else {
				htmlElement.setAttribute('lang', 'en');
			}
		}

		return;
	};

	connectedCallback() {
		this.updateLanguage();
	}

	render() {
		const language = this.language === 'en' ? 'Fran√ßais' : 'English';
		const abbreviatedLanguage = this.language === 'en' ? 'FR' : 'EN';

		return (
			<a
				class={
					this.size === 'default'
						? 'ontario-language-toggler ontario-language-toggler--default'
						: 'ontario-language-toggler ontario-language-toggler--small'
				}
				href={this.url ? this.url : '#'}
				aria-label={this.translations.languageToggle.ariaLabel[`${this.language}`]}
				onClick={(e) => this.handleHeaderLanguageToggled(this.language, e)}
			>
				{this.size === 'small' ? (
					<span>{language}</span>
				) : (
					<Fragment>
						<abbr title={language} class="ontario-show-for-small-only">
							{abbreviatedLanguage}
						</abbr>
						<span class="ontario-show-for-medium">{language}</span>
					</Fragment>
				)}
			</a>
		);
	}
}
