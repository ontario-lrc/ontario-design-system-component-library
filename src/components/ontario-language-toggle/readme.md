# ontario-language-toggle

This component is used internally to toggle the component languages, set translations and update the HTML lang attribute in the DOM.

It renders an anchor tag (styles as an ontario-button) that accepts either a URL prop to switch the language to, or a custom toggle function if additional functionality is necessary.

It is used in the ontario-header component.

<!-- Auto Generated Below -->

## Properties

| Property               | Attribute  | Description                                                                                         | Type                                | Default     |
| ---------------------- | ---------- | --------------------------------------------------------------------------------------------------- | ----------------------------------- | ----------- |
| `customLanguageToggle` | --         | A custom function to pass to the language toggle button. This is optional.                          | `Function \| undefined`             | `undefined` |
| `language`             | `language` |                                                                                                     | `string`                            | `undefined` |
| `size`                 | `size`     | The size of the language toggle button. If no prop is passed, it will be set to the `default` size. | `"default" \| "small" \| undefined` | `'default'` |
| `url`                  | `url`      | The URL to change to when the language toggle button is clicked. This is optional.                  | `string \| undefined`               | `undefined` |

## Events

| Event                   | Description                                                                                           | Type                  |
| ----------------------- | ----------------------------------------------------------------------------------------------------- | --------------------- |
| `headerLanguageToggled` | An event that emits to other components that the language toggle button has been toggled.             | `CustomEvent<string>` |
| `setAppLanguage`        | An event to set the Document's HTML lang property, and emit the toggled language to other components. | `CustomEvent<string>` |

## Dependencies

### Used by

- [ontario-header](../ontario-header)

### Graph

```mermaid
graph TD;
  ontario-header --> ontario-language-toggle
  style ontario-language-toggle fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
