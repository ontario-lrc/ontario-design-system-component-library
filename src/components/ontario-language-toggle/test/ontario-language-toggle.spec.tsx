import { newSpecPage } from '@stencil/core/testing';
import { OntarioLanguageToggle } from '../ontario-language-toggle';

describe('ontario-language-toggle', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioLanguageToggle],
				html: `<ontario-language-toggle></ontario-language-toggle>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});

	it('should render the default size language toggle if no size is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioLanguageToggle],
			html: `<ontario-language-toggle></ontario-language-toggle>`,
		});

		expect(page.root).toEqualHtml(`
			<ontario-language-toggle>
				<mock:shadow-root>
					<a aria-label="Click to switch language to French" class="ontario-language-toggler ontario-language-toggler--default" href="#">
						<abbr class="ontario-show-for-small-only" title="Français">
							FR
						</abbr>
						<span class="ontario-show-for-medium">
							Français
						</span>
					</a>
				</mock:shadow-root>
			</ontario-language-toggle>
		`);
		expect(page.rootInstance.size).toBe('default');
	});

	it('should render a default language of English if no language prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioLanguageToggle],
			html: `<ontario-language-toggle></ontario-language-toggle>`,
		});

		expect(page.rootInstance.language).toBe('en');
	});

	it('should render a small size language toggle with a French default language when explicitly specified', async () => {
		const page = await newSpecPage({
			components: [OntarioLanguageToggle],
			html: `<ontario-language-toggle size="small" url="/en" language="fr"></ontario-language-toggle>`,
		});

		expect(page.root).toEqualHtml(`
			<ontario-language-toggle size="small" url="/en" language="fr">
				<mock:shadow-root>
					<a aria-label="Cliquez pour changer de langue en anglais" class="ontario-language-toggler ontario-language-toggler--small" href="/en">
						<span>
							English
						</span>
					</a>
				</mock:shadow-root>
			</ontario-language-toggle>
		`);
	});
});
