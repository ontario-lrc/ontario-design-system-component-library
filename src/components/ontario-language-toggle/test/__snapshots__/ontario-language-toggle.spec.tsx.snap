// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-language-toggle snapshot should render the expected html 1`] = `
<ontario-language-toggle>
  <mock:shadow-root>
    <a aria-label="Click to switch language to French" class="ontario-language-toggler ontario-language-toggler--default" href="#">
      <abbr class="ontario-show-for-small-only" title="Français">
        FR
      </abbr>
      <span class="ontario-show-for-medium">
        Français
      </span>
    </a>
  </mock:shadow-root>
</ontario-language-toggle>
`;
